File: scripts/generate-status-snippets.mjs
#!/usr/bin/env node

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const repoRoot = path.resolve(__dirname, '..');
const statusPath = path.join(repoRoot, 'status', 'feature-readiness.json');
const generatedDir = path.join(repoRoot, 'status', 'generated');
const featureTablePath = path.join(generatedDir, 'feature-table.md');
const phaseSummaryPath = path.join(generatedDir, 'phase-summary.md');

function loadReadiness() {
  return JSON.parse(fs.readFileSync(statusPath, 'utf8'));
}

function sortFeatureEntries(featuresObj) {
  return Object.entries(featuresObj).sort(([a], [b]) => Number(a) - Number(b));
}

export function renderFeatureTable(readiness) {
  const rows = sortFeatureEntries(readiness.features).map(([id, feature]) => {
    return `| \`${id}\` | ${feature.friendly_name} | \`${feature.lifecycle_state}\` | \`${feature.implementation_state}\` | \`${feature.production_readiness}\` |`;
  });

  return [
    '<!-- GENERATED: status/feature-readiness.json -->',
    '| Feature | Name | Lifecycle | Implementation | Production |',
    '|---|---|---|---|---|',
    ...rows,
    '',
  ].join('\n');
}

export function renderPhaseSummary(readiness) {
  const entries = sortFeatureEntries(readiness.features);
  const lifecycleCounts = {};
  const productionCounts = {};

  for (const [, feature] of entries) {
    lifecycleCounts[feature.lifecycle_state] = (lifecycleCounts[feature.lifecycle_state] ?? 0) + 1;
    productionCounts[feature.production_readiness] = (productionCounts[feature.production_readiness] ?? 0) + 1;
  }

  const lifecycleLines = Object.entries(lifecycleCounts)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([state, count]) => `- ${state}: ${count}`);
  const productionLines = Object.entries(productionCounts)
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([state, count]) => `- ${state}: ${count}`);

  return [
    '<!-- GENERATED: status/feature-readiness.json -->',
    `Updated: ${readiness.updated_at}`,
    '',
    'Lifecycle counts:',
    ...lifecycleLines,
    '',
    'Production-readiness counts:',
    ...productionLines,
    '',
  ].join('\n');
}

export function writeGeneratedSnippets(readiness) {
  fs.mkdirSync(generatedDir, { recursive: true });
  fs.writeFileSync(featureTablePath, renderFeatureTable(readiness), 'utf8');
  fs.writeFileSync(phaseSummaryPath, renderPhaseSummary(readiness), 'utf8');
}

function runCli() {
  const readiness = loadReadiness();
  writeGeneratedSnippets(readiness);
  process.stdout.write(
    [
      'Generated status snippets:',
      `- ${path.relative(repoRoot, featureTablePath)}`,
      `- ${path.relative(repoRoot, phaseSummaryPath)}`,
      '',
    ].join('\n'),
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  runCli();
}

File: scripts/generate-status-snippets.py
#!/usr/bin/env python3
"""Generate status markdown snippets from status/feature-readiness.json.

Usage:
  python scripts/generate-status-snippets.py         # write files
  python scripts/generate-status-snippets.py --check # fail if out of date
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
STATUS_FILE = ROOT / "status" / "feature-readiness.json"
OUTPUT_FILE = ROOT / "status" / "generated" / "feature-table.md"


def load_status() -> dict:
    return json.loads(STATUS_FILE.read_text())


def render_table(status: dict) -> str:
    rows = []
    features: dict[str, dict] = status["features"]
    for fid in sorted(features.keys()):
        item = features[fid]
        rows.append(
            f"| {fid} | {item['lifecycle_state']} | {item['implementation_state']} | "
            f"{item['production_readiness']} | {item['notes']} |"
        )

    header = [
        "# Generated Feature Readiness",
        "",
        f"Source: `status/feature-readiness.json` (updated_at: {status['updated_at']})",
        "",
        "| Feature | Lifecycle | Implementation | Readiness | Notes |",
        "|---|---|---|---|---|",
    ]

    return "\n".join(header + rows) + "\n"


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--check", action="store_true", help="verify generated files are up to date")
    args = parser.parse_args()

    if not STATUS_FILE.exists():
        print(f"ERROR: missing status file: {STATUS_FILE}", file=sys.stderr)
        return 1

    status = load_status()
    content = render_table(status)

    if args.check:
        if not OUTPUT_FILE.exists():
            print(f"ERROR: missing generated file: {OUTPUT_FILE}", file=sys.stderr)
            return 1
        current = OUTPUT_FILE.read_text()
        if current != content:
            print("ERROR: generated status snippets are out of date", file=sys.stderr)
            return 1
        print("Generated status snippets are up to date")
        return 0

    OUTPUT_FILE.parent.mkdir(parents=True, exist_ok=True)
    OUTPUT_FILE.write_text(content)
    print(f"Wrote {OUTPUT_FILE}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
